stablehlo/tests/CheckOps.cpp

```diff
@@ -156,6 +157,46 @@ llvm::Error evalExpectSerializedEqOp(const Tensor &expected, StringRef probeId,
   return evalExpectEqOp(expected, *tensor);
 }
 
+inline std::string toString(const Element &elem) {
+  auto type = elem.getType();
+  if (isSupportedFloatType(type)) {
+    SmallString<40> S;
+    elem.getFloatValue().toString(S);
+    return std::string(S);
+  } else if (isSupportedComplexType(type)) {
+    auto c = elem.getComplexValue();
+    SmallString<80> S("(");
+    c.real().toString(S);
+    S += ", ";
+    c.imag().toString(S);
+    S += ")";
+    return std::string(S);
+  } else {
+    return debugString(elem);
+  }
+}
+
+llvm::Error evalExpectIsCloseOp(const Tensor &lhs, const Tensor &rhs,
+                                const Tensor &abs_error, const Tensor &input) {
```

Comment:
What is the relevance of `input` here given that it is used only for reporting error? IOW,

Reply:
The `input` is the input to `asin` operator, `lhs` is the value of `asin(input)`, `rhs` is the expected value. That said, when the is-close test fails for `lhs` and `rhs`, one would need to know which `input` this failure corresponds to. This is required to determine which part of the asin implementation is responsible for producing the incorrect result.

For example, when disabling the implemented algorithm, a part of a failure message from running `stablehlo/tests/math/asin_complex64.mlir`, is:
```
LLVM ERROR: Error evaluating function: main. 
	Fallback for check.expect_is_close failed: Elements values don't match:
...
  index=[54], actual=(NaN, NaN), expected=(-1.57079637, -89.4159851), atol=2.42017668E-5, input=(-3.40282326E+38, -1.53329339E-19)
  index=[56], actual=(-1.53329339E-19, -5.87747175E-39), expected=(-1.53329339E-19, -1.53329339E-19), atol=4.75334363E-26, input=(-1.53329339E-19, -1.53329339E-19)
  index=[57], actual=(-1.17549435E-38, 0), expected=(-1.17549435E-38, -1.53329339E-19), atol=4.11823295E-26, input=(-1.17549435E-38, -1.53329339E-19)
...
```
While it is true that the `index` and `input` parts have one-to-one relation, it is just a very tedious task to find the `input` value from `index` (it requires decoding `samples` data in asin_complex64.mlir) while it is much more convenient to read the `input` value from the failure message and confirm that, indeed, the actual value, say, `(-1.5E-19, -5.8E-39)` is incorrect for `asin((-1.5E-19, -1.5E-19))`.

If `input` is not shown, then having the information such as `index=[56], actual=(-1.5E-19, -5.8E-39), expected=(-1.5E-19, -1.5E-19)`, is not very useful to determine what-is-wrong/where-to-look-for-a-fix until one determines the `input` value as well.

Reply:
>> That said, when the is-close test fails for lhs and rhs, one would need to know which input this failure corresponds to. This is required to determine which part of the asin implementation is responsible for producing the incorrect result.

I agree to the overall setiment towards using the input value. However, assuming that `evalExpectIsCloseOp` is intended to be used more generally, how would users decide which function to apply to the input? 

Also, what if the `lhs` is generated by a sequence of operations before they are compared using `expect_is_close`. In that case the questions would be
(1) What is the input? The one after the previous `expect_is_close` operation?
(2) How would the user know which function to apply to this input for debuggability. 

IMO it may not be a good user experience for a user of the check* operation to care about passing the original `input` the `lhs` is generated from plus the function to transform `input` to `lhs` (in some format to capture the case of multiple operations transforming `input` to `lhs`).

wdyt?

Reply:
In case you agree to my point we can create an overload of `check.expect_almost_equal` op with tolerances specified by user and its eval function can resuse `evalExpectAlmostEqOp` ?

Reply:
On a second thought: why do we need separate tolerances for individual elements of a tensor? `asin` being an elementwise all the elements of the tensor are supposed to be generated using a unified algorithm. Or is this something specific to Hull et. al. algorithm?

My main concern here is  the lines of user experience, for example, how these individual tolerances are to be generated?

Reply:
> On a second thought: why do we need separate tolerances for individual elements of a tensor? `asin` being an elementwise all the elements of the tensor are supposed to be generated using a unified algorithm. Or is this something specific to Hull et. al. algorithm?

The non-zero finite inputs to, say, fp64 version of `asin`, can range from `2.2250738585072014e-308` to `1.7976931348623157e+308`. The absolute values of complex `asin` ranges from `2.2250738585072014e-308` to approximately `710`. Recall, a proper value for `atol` depends on the magnitude of the numbers being compared. For instance, when using `atol=1e-15` for comparing `asin` values that are, say, larger than 1, is appropriate then using the same `atol` value for comparing `asin` values in the order of, say, `1e-100`, is totally useless because `abs(f - g) < 1e-15` always holds.

In the case of `asin`, we have an upper bound to `asin` values. In the case of other math functions, the range of values may be from `1e-308` to `1e+308` and one must choose `atol` value appropriately for sensible testing.

> My main concern here is the lines of user experience, for example, how these individual tolerances are to be generated?

I think there is a confusion of context here. Testing of `asin` correctness on the whole complex plane has nothing to do with an user experience other than stablehlo being able to declare to users that `asin` evaluation for any complex input produces mathematically correct result within the restrictions of the used floating point system. Users calling `asin` do not need to deal with specifying the tolerances at all. The provided Python script that generates these asin tests take care of computing the tolerances appropriately.

Reply:
> I agree to the overall setiment towards using the input value. However, assuming that evalExpectIsCloseOp is intended to be used more generally, how would users decide which function to apply to the input?

The assumption of using evalExpectIsCloseOp more generally is OT for this PR. However, I can see that "is close" may be confused with a more general problem of comparing two arbitrary values while here we are dealing with values that (i) are approximations (e.g. the value of `stablhlo.asin(input)`) and (ii) are references (e.g. the value of `mpmath.asin(input)` that is taken as the mathematical truth as it is evaluated using multi-precision fp arithmetic).

To be more explicit about the intention of the used method, I suggest replacing "expect_is_close" with "validate_against" or "expect_close_to_reference" or similar:
```
expect_close_to_reference(actual, reference, abs_allowed_error, input)
```

What do you think?

Reply:
The recent update to this PR introduces:
```
check.expect_close_to_reference(actual, reference, input, max_ulp_difference=1)
```
that is available only within `stablehlo/tests/` and it uses ULP difference between actual and reference values to measure the closeness. This eliminated the need to introduce `abs_allowed_error` tensor argument and is actually a much better measure for closeness as it does not depend on the magnitudes of the values.

Still keeping the `input` argument as it is very useful when developing algorithms such as for `asin`.